src/gleam/erlang/process.gleam

```diff
@@ -64,19 +70,72 @@ fn spawn_link(a: fn() -> anything) -> Pid
 ///
 pub opaque type Subject(message) {
   Subject(owner: Pid, tag: Reference)
+  NamedSubject(name: Name(message))
+}
+
+/// A name is an identity that a process can adopt, after which they will receive
+/// messages sent to that name. This has two main advantages:
+///
+/// - Structuring OTP programs becomes easier as a name can be passed down the
+///   program from the top level, while without names subjects and pids would
+///   need to be passed up from the started process and then back down to the
+///   code that works with that process.
+/// - A new process can adopt the name of one that previously failed, allowing
+///   it to transparently take-over and handle messages that are sent to that
+///   name.
+///
+/// Names are globally unique as each process can have at most 1 name, and each
+/// name can be registered by at most 1 process. Create all the names your
+/// program needs at the start of your program and pass them down. Names are
+/// Erlang atoms internally, so never create them dynamically. Generating too
+/// many atoms will result in the atom table getting filled and causing the entire
+/// virtual machine to crash.
+///
+/// The most commonly used name functions are `new_name`, `register`, and
+/// `named_subject`.
+///
+pub type Name(message)
+
+/// Generate a new name that a process can register itself with using the
+/// `register` function, and other processes can send messages to using
+/// `named_subject`.
+///
+/// ## Safe use
+///
+/// Use this function to create all the names your program needs when it
+/// starts. **Never call this function dynamically** such as within a loop or
+/// within a process within a supervision tree.
+///
+/// Each time this function is called a new atom will be generated. Generating
+/// too many atoms will result in the atom table getting filled and causing the
+/// entire virtual machine to crash.
+///
+@external(erlang, "gleam_erlang_ffi", "new_name")
+pub fn new_name() -> Name(message)
+
+/// Create a subject for a name, which can be used to send and receive messages.
+///
+pub fn named_subject(name: Name(message)) -> Subject(message) {
+  NamedSubject(name)
```

Comment:
With reservation for not completely understanding this, it looks as if you only can have one Subject for a named process. How would you create different subjects for a named process?
The new_subject creates a new reference each time you call it, which means you may have several Subjects (Message sets).

Reply:
A process can have as many unnamed subjects as they want (useful for ad-hoc messages such as the reply in `call`), but each process can only have one name, and the API of that name is a single message type. To me this seemed in keeping with how Erlang names are used.

Perhaps I am making a mistake here and this is too restrictive. Do you have thoughts on how one may want to use multiple distinct subjects for one name?

Reply:
OK, thanks.
That makes sense, and references in tags will not survive a restart, so you would have to re-register those as well as the process id. I guess that if you have a service that present interfaces for different types of clients (using different protocols) this could be of interest, but maybe you just solve it by presenting Subject handles to the different interfaces from the "single API".